File: clients/typescript/switchkeys/core/exceptions.ts
import { SwitchKeysLogger } from "../utils/logger";

const logger: SwitchKeysLogger = new SwitchKeysLogger();

/**
 * Custom error class for representing connection errors in SwitchKeys.
 */
class SwitchKeysConnectionError extends Error {
  /**
   * Constructs a new SwitchKeysConnectionError with the specified error message.
   * @param message The error message.
   */
  constructor(message: string) {
    super(message);
    logger.error(message);
  }
}

/**
 * Custom error class for representing database record not found errors in SwitchKeys.
 */
class SwitchKeysRecordNotFoundError extends Error {
  /**
   * Constructs a new SwitchKeysConnectionError with the specified error message.
   * @param message The error message.
   */
  constructor(message: string) {
    super(message);
    logger.error(message);
  }
}

export { SwitchKeysConnectionError, SwitchKeysRecordNotFoundError };

File: clients/dart/switchkeys/lib/src/core/exceptions.dart
import 'package:switchkeys/src/utils/logger.dart';

/// Base class for all kinds of client-side errors.
class Error implements Exception {
  @override
  String toString() {
    return runtimeType.toString();
  }
}

/// Indicates connection error.
class SwitchKeysConnectionError extends Error {
  final String message;

  SwitchKeysConnectionError(this.message);

  @override
  String toString() {
    return message;
  }
}

/// Indicates connection error.
class SwitchKeysRecordNotFoundError extends Error {
  final String message;

  SwitchKeysRecordNotFoundError(this.message);

  @override
  String toString() {
    return message;
  }
}

/// Indicates a broken response package.
class ResponseError extends Error {
  final String message;

  ResponseError(this.message);

  @override
  String toString() {
    return message;
  }
}

/// Indicates a missing authentication.
class AuthenticationError extends Error {
  AuthenticationError() {
    SwitchKeysLogger().error(
        'There are no tokens found in the config file, and none were specified when initializing an instance of the SwitchKeys client. Please ensure that tokens are provided either in the config file or passed during initialization.');
  }
}

/// Indicates that a feature is not enabled.
class FeatureNotEnabled implements Exception {
  final String message;
  final dynamic errors;

  FeatureNotEnabled(this.message, {this.errors});

  @override
  String toString() {
    return 'FeatureNotEnabled: $message';
  }
}
